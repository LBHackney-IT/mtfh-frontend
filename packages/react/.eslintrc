{
  "extends": "@lbh-mtfh/eslint-config",
  "parserOptions": {
    "project": "tsconfig.json"
  },
  "rules": {
    "prefer-arrow-callback": "off",
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*/rollup.config.js"]
      }
    ],
    "jsx-a11y/anchor-is-valid": [
      "error",
      {
        "components": ["Link"],
        "aspects": ["invalidHref"]
      }
    ]
  }
}
