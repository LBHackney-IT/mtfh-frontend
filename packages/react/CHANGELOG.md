# @lbh-mtfh/react

## 1.1.0

### Minor Changes

- 37b7b3e: Initial release

### Patch Changes

- Updated dependencies [37b7b3e]
  - @lbh-mtfh/system@1.1.0
  - @lbh-mtfh/test-utils@1.1.0
