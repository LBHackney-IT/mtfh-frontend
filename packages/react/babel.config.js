const babel = require("../../babel.config")

module.exports = babel
