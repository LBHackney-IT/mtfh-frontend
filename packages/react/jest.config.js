const baseConfig = require("../../jest.config")

baseConfig.setupFilesAfterEnv.push("<rootDir>/jest.setup.ts")

module.exports = {
  ...baseConfig,
}
