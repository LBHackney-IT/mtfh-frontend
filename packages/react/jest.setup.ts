import * as LBHFrontend from "lbh-frontend"

window.LBHFrontend = LBHFrontend
