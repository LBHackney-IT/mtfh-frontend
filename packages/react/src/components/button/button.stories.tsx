import React from "react";

import { Button } from "./button";

export default {
  title: "Button",
  component: Button,
};

export const Base = () => {
  return <Button>Button</Button>;
};

export const Secondary = () => {
  return <Button variant="secondary">Secondary Button</Button>;
};

export const ButtonLink = () => {
  return (
    <Button as="a" href="/">
      Button Link
    </Button>
  );
};

export const LoadingIndicator = () => {
  return <Button isLoading>Button</Button>;
};

export const LoadingIndicatorWithText = () => {
  return (
    <Button loadingText="I'm Loading" isLoading>
      Button
    </Button>
  );
};
