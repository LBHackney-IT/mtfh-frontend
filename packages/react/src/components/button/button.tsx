import React, { forwardRef } from "react";

import cn from "classnames";

import { widthOverrides } from "../../utils";
import styles from "./styles.module.scss";

import type * as Polymorphic from "@radix-ui/react-polymorphic";

export interface ButtonProps {
  variant?: "primary" | "secondary";
  isLoading?: boolean;
  isDisabled?: boolean;
  loadingText?: string;
  override?: number;
}

export type ButtonComponent = Polymorphic.ForwardRefComponent<"button", ButtonProps>;

export const Button: ButtonComponent = forwardRef(function Button(
  {
    as: ButtonComp = "button",
    variant = "primary",
    isLoading = false,
    loadingText,
    isDisabled,
    children,
    className,
    override,
    ...props
  },
  ref,
) {
  const buttonClasses = cn(
    styles.govukButton,
    styles.lbhButton,
    {
      [`${styles.govukButtonSecondary} ${styles.lbhButtonSecondary}`]:
        variant === "secondary",
      [`${styles.govukButtonDisabled} ${styles.lbhButtonDisabled}`]: isDisabled,
    },
    widthOverrides(override),
    className,
  );

  const disabled = isDisabled || isLoading || undefined;

  return (
    <ButtonComp
      ref={ref}
      className={buttonClasses}
      type={ButtonComp === "button" ? "button" : undefined}
      disabled={ButtonComp === "button" ? disabled : undefined}
      aria-disabled={disabled}
      {...props}
    >
      {isLoading && (
        <span className={styles.buttonLoadingIndicator}>
          <span>Loading...</span>
        </span>
      )}
      {isLoading && loadingText ? loadingText : children}
    </ButtonComp>
  );
});
