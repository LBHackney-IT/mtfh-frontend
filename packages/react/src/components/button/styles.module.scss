@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-button/button";

.button-loading-indicator {
  display: inline-block;
  border-top: 2px solid currentcolor;
  border-right: 2px solid currentcolor;
  border-bottom-style: solid;
  border-left-style: solid;
  border-radius: 50%;
  border-bottom-width: 2px;
  border-left-width: 2px;
  border-bottom-color: transparent;
  border-left-color: transparent;
  width: 1em;
  height: 1em;
  color: currentColor;
  margin-right: 1em;
  vertical-align: middle;
  animation: buttonLoadingIndicatorRotation 0.45s linear 0s infinite normal none running;

  span {
    border: 0px;
    clip: rect(0px, 0px, 0px, 0px);
    height: 1px;
    width: 1px;
    margin: -1px;
    padding: 0px;
    overflow: hidden;
    white-space: nowrap;
    position: absolute;
  }
}

@keyframes buttonLoadingIndicatorRotation {
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
}
