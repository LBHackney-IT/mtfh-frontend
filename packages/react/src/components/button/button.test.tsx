import React, { useEffect } from "react";
import { render, screen } from "@testing-library/react";

import { testA11y } from "@lbh-mtfh/test-utils";
import { Button } from "./button";

test("it renders correctly", async () => {
  const { container } = render(<Button>Test Button</Button>);
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it shows a loading state", async () => {
  const { container } = render(<Button isLoading>Test Button</Button>);
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it shows a loading state with custom text", () => {
  render(
    <Button isLoading loadingText="Processing">
      Test Button
    </Button>
  );
  expect(screen.getByText("Processing")).toBeInTheDocument();
});

test("it renders the correct variant", async () => {
  const { container } = render(<Button variant="secondary">Test Link</Button>);
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it accepts a ref", () => {
  const callback = jest.fn();
  const Comp = () => {
    const ref = React.createRef<HTMLButtonElement>();

    useEffect(() => {
      if (ref.current) {
        callback(ref.current);
      }
    }, [ref]);
    return <Button ref={ref}>Test Button</Button>;
  };
  render(<Comp />);
  expect(callback).toBeCalledTimes(1);
});
