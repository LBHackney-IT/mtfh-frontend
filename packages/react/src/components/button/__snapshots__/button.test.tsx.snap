// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<div>
  <button
    class="govukButton lbhButton"
    type="button"
  >
    Test Button
  </button>
</div>
`;

exports[`it renders the correct variant 1`] = `
<div>
  <button
    class="govukButton lbhButton govukButtonSecondary lbhButtonSecondary"
    type="button"
  >
    Test Link
  </button>
</div>
`;

exports[`it shows a loading state 1`] = `
<div>
  <button
    aria-disabled="true"
    class="govukButton lbhButton"
    disabled=""
    type="button"
  >
    <span
      class="buttonLoadingIndicator"
    >
      <span>
        Loading...
      </span>
    </span>
    Test Button
  </button>
</div>
`;
